hackathonbaobab2021/core/experiment.py

```diff
@@ -59,33 +72,418 @@ def from_zipped_json(
+        """
         assignment = self.solution.get_assignment()
         teams = self.instance.get_teams("id").vapply(lambda v: 0)
-        num_matches = (len(teams) - 1) * 2
+        num_matches = len(teams) - 1
         return (
-            teams
-            + assignment.to_dict(result_col=["slot"], indices=["home"]).to_lendict()
-        ).vfilter(lambda v: v != num_matches)
+            assignment.to_dict(result_col=["slot"], indices=[mode])
+            .to_lendict()
+            .fill_with_default(teams)
+            .vfilter(lambda v: v != num_matches)
+        )
+
+    def get_objective(self, **params) -> float:
+        # for each check, the violation value
+        checks = self.check_solution(c_type=SOFT).to_dictdict().to_dictup()
+        # CA4 has CA4_global and CA4_slot so we need to get the name by looking for _
+        get_name = lambda name: name.split("_")[0]
+        # the first key is the constraint tag, the second is the constraint id.
+        get_penaly = lambda k: self.instance.get_penalty(get_name(k[0]), k[1])
+        # we get the absolute deviation, then
+        # we multiply the violation with the penalty, then sum the results
+        return sum(checks.kvapply(lambda k, v: abs(v) * get_penaly(k)).values())
+
+    def to_xml(self, path, instance_name="Test Instance Demo"):
+        root = ET.Element("Solution")
+        metadata = self._xml_build_metadata(instance_name)
+        games = self._xml_build_games()
+        root.append(metadata)
+        root.append(games)
+        tree = ET.ElementTree(element=root)
+        # this following function gives the indentation and spacing:
+        indent(root)
+        tree.write(
+            path, xml_declaration=True, encoding="utf-8", short_empty_elements=False
+        )
+
+    def _xml_build_metadata(self, instance_name):
+        metadata = ET.Element("MetaData")
+        solution = ET.Element("SolutionName")
+        solution.text = "Sol" + instance_name
+        instance = ET.Element("InstanceName")
+        instance.text = instance_name
+        objective = ET.Element("ObjectiveValue")
+        objective_contents = dict(
+            infeasibility=str(sum(self.check_solution().to_lendict().values())),
+            objective=str(self.get_objective()),
+        )
+        for v in objective_contents.items():
+            objective.set(*v)
+        for el in [instance, solution, objective]:
+            metadata.append(el)
+        return metadata
+
+    def _xml_build_games(self):
+        games = ET.Element("Games")
+
+        def to_et(value):
+            el = ET.Element("ScheduledMatch")
+            for v in value.items():
+                el.set(*v)
+            return el
+
+        assignment = self.solution.get_assignment().vapply(to_et)
+        for el in assignment:
+            games.append(el)
+        return games
+
+
+def compare(_dict: SuperDict[Any, int], constraints, side: Union[str, None] = "min"):
```

Comment:
Could the methods that are define here onwards be included in a `tools.py` file in the core folder?

Reply:
they are very specific to the experiment... I guess that's why I have them there. I could move them. They could have been static methods but I think static methods take too much space inside the class.

Reply:
Ok, no problem